File: kata/bowling_game/c/2024_01_20/tests/test_bowling_game.c
#include <CUnit/Basic.h>
#include <CUnit/CUnit.h>

#include <stdlib.h>

#include "../src/game.h"

Game *game;

void set_up() { game = new_game(); }

void roll_many(int n, int pins) {
  for (int i = 0; i < n; i++) {
    game->roll(game, pins);
  }
}

void roll_standard_spare() {
  game->roll(game, 5);
  game->roll(game, 5);
}

void roll_different_spre() {
  game->roll(game, 2);
  game->roll(game, 8);
}

void roll_strike() {
  game->roll(game, 10);
}

void test_all_gutters(void) {
  set_up();
  roll_many(20, 0);

  CU_ASSERT_EQUAL(0, game->get_score(game));
}

void test_all_ones(void) {
  set_up();
  roll_many(20, 1);

  CU_ASSERT_EQUAL(20, game->get_score(game));
}

void test_standard_one_spare(void) {
  set_up();
  roll_standard_spare();
  game->roll(game, 3);
  roll_many(17, 0);

  CU_ASSERT_EQUAL(16, game->get_score(game));
}

void test_different_one_spare(void) {
  set_up();
  roll_different_spre();
  game->roll(game, 3);
  roll_many(17, 0);

  CU_ASSERT_EQUAL(16, game->get_score(game));
}

void test_one_strike(void) {
  set_up();
  roll_strike();
  game->roll(game, 3);
  game->roll(game, 4);
  roll_many(16, 0);

  CU_ASSERT_EQUAL(24, game->get_score(game));
}

void test_perfect_game(void) {
  set_up();
  roll_many(12, 10);

  CU_ASSERT_EQUAL(game->get_score(game), 300);
}


int main() {
  CU_initialize_registry();
  CU_pSuite suite = CU_add_suite("BowlingGameTest", NULL, NULL);

  if (NULL != suite) {
    CU_add_test(suite, "test of all gutters", test_all_gutters);
    CU_add_test(suite, "test of all ones", test_all_ones);
    CU_add_test(suite, "test of standard spare", test_standard_one_spare);
    CU_add_test(suite, "test of different spare", test_different_one_spare);
    CU_add_test(suite, "test of one strike", test_one_strike);
    CU_add_test(suite, "test of perfect game", test_perfect_game);
  }

  CU_basic_set_mode(CU_BRM_VERBOSE);
  CU_basic_run_tests();
  CU_cleanup_registry();

  return CU_get_error();
}

File: kata/bowling_game/python/2025_07_08/test_bowling_game.py
import unittest

from bowling_game import Game

class BowlingGameTest(unittest.TestCase):
    def setUp(self):
        self.game = Game()

    def test_gutter_game(self):
        self._roll_many(20, 0)
        self.assertEqual(0, self.game.score())

    def test_all_one(self):
        self._roll_many(20, 1)
        self.assertEqual(20, self.game.score())

    def test_one_spare(self):
        self.game.roll(5)
        self.game.roll(5)
        self.game.roll(3)
        self._roll_many(17, 0)
        self.assertEqual(16, self.game.score())

    def test_one_strike(self):
        self.game.roll(10)
        self.game.roll(3)
        self.game.roll(4)
        self._roll_many(16, 0)
        self.assertEqual(24, self.game.score())

    def _roll_many(self, n ,pins):
        for _ in range(n):
            self.game.roll(pins)
